### PR TITLE
GH-707: Introduced the `@theia/typehierarchy` extension.

### PR DESCRIPTION
**Pull request by kittaakos**
- Extended the LSP with the type hierarchy.
 - From now on, a tree node can carry decoration data.
 - Added editor access for the current/active editors.
 - Wired the JDT LS type hierarchy into the application.

Closes: #707.

Signed-off-by: Akos Kitta <kittaakos@typefox.io>

Ready for preview:
screencast 2018-10-11 15-18-32

### CODE REVIEW

**Comment by AlexTugarev** [comment]
screen shot 2018-10-11 at 15 55 08

just started testing... looks awesome!

**Comment by kittaakos** [comment]
When restarting the Theia backend without refreshing the browser:
```
root ERROR RangeError: Maximum call stack size exceeded
    at http://localhost:3000/11.bundle.js:3892:19
    at Map.forEach (<anonymous>)
    at SubTypeHierarchyFeature.push.../../node_modules/vscode-base-languageclient/lib/client.js.TextDocumentFeature.dispose (http://localhost:3000/11.bundle.js:3891:25)
    at SubTypeHierarchyFeature.push.../../packages/languages/lib/browser/typehierarchy/typehierarchy-feature.js.TypeHierarchyFeature.dispose (http://localhost:3000/36.bundle.js:381:34)
    at http://localhost:3000/36.bundle.js:628:42
    at CallbackList.../../packages/core/lib/common/event.js.CallbackList.invoke (http://localhost:3000/bundle.js:100223:39)
    at Emitter.../../packages/core/lib/common/event.js.Emitter.fire (http://localhost:3000/bundle.js:100303:36)
    at SubTypeHierarchyFeature.push.../../packages/languages/lib/browser/typehierarchy/typehierarchy-feature.js.TypeHierarchyFeature.dispose (http://localhost:3000/36.bundle.js:380:32)
    at http://localhost:3000/36.bundle.js:628:42
    at CallbackList.../../packages/core/lib/common/event.js.CallbackList.invoke (http://localhost:3000/bundle.js:100223:39)
root ERROR RangeError: Maximum call stack size exceeded
    at http://localhost:3000/11.bundle.js:3892:19
    at Map.forEach (<anonymous>)
    at SubTypeHierarchyFeature.push.../../node_modules/vscode-base-languageclient/lib/client.js.TextDocumentFeature.dispose (http://localhost:3000/11.bundle.js:3891:25)
    at SubTypeHierarchyFeature.push.../../packages/languages/lib/browser/typehierarchy/typehierarchy-feature.js.TypeHierarchyFeature.dispose (http://localhost:3000/36.bundle.js:381:34)
    at http://localhost:3000/36.bundle.js:628:42
    at CallbackList.../../packages/core/lib/common/event.js.CallbackList.invoke (http://localhost:3000/bundle.js:100223:39)
    at Emitter.../../packages/core/lib/common/event.js.Emitter.fire (http://localhost:3000/bundle.js:100303:36)
    at SubTypeHierarchyFeature.push.../../packages/languages/lib/browser/typehierarchy/typehierarchy-feature.js.TypeHierarchyFeature.dispose (http://localhost:3000/36.bundle.js:380:32)
    at http://localhost:3000/36.bundle.js:628:42
    at CallbackList.../../packages/core/lib/common/event.js.CallbackList.invoke (http://localhost:3000/bundle.js:100223:39)
```
Possible related: #3150, #3145. Something is odd with the reconnection in Java.

File: packages/typehierarchy/src/browser/tree/typehierarchy-tree-container.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import { TypeHierarchyTree } from './typehierarchy-tree';
+import { TypeHierarchyTreeModel } from './typehierarchy-tree-model';
+import { TypeHierarchyTreeWidget } from './typehierarchy-tree-widget';
+
+function createHierarchyTreeContainer(parent: interfaces.Container): Container {
+    const child = createTreeContainer(parent);
+
+    child.unbind(TreeImpl);
+    child.bind(TypeHierarchyTree).toSelf();
+    child.rebind(Tree).toDynamicValue(ctx => ctx.container.get(TypeHierarchyTree));
```
Comment (kittaakos) [inline packages/typehierarchy/src/browser/tree/typehierarchy-tree-container.ts]: TODO: https://github.com/theia-ide/theia/pull/3153

`toDynamicValue` -> `toService`.

**Comment by simark** [comment]
Could you point me to the latest version of the LSP proposal for type hierarchy?  I would like to test this with clangd, if I have the time.

**Comment by kittaakos** [comment]
> latest version of the LSP proposal for type hierarchy

https://github.com/Microsoft/vscode-languageserver-node/pull/346

Note, this implementation differs from the proposed one.

----

Extracted from the source:
 - [Client capabilities](https://github.com/theia-ide/theia/pull/3148/files#diff-1c813b550f66b5a56ebec5bc0f0c5f1dR61)
 - [Server capabilities](https://github.com/theia-ide/theia/pull/3148/files#diff-1c813b550f66b5a56ebec5bc0f0c5f1dR70)
 - Language features
   - [`textDocument/subTypes`](https://github.com/theia-ide/theia/pull/3148/files#diff-f25ee20dffce3fe36e4f9a0394edec16R60)
   - [`textDocument/superTypes`](https://github.com/theia-ide/theia/pull/3148/files#diff-f25ee20dffce3fe36e4f9a0394edec16R68)
 - Added [optional `uri?: string`](https://github.com/theia-ide/theia/pull/3148/files#diff-f25ee20dffce3fe36e4f9a0394edec16R34) field to the `DocumentSymbol`.

Or in Java/Xtend: https://github.com/eclipse/lsp4j/pull/273

**Comment by kittaakos** [comment]
I have noticed another UX issue, we **should not** reveal the document symbol, when expanding the nodes in the type hierarchy tree.

**Comment by AlexTugarev** [comment]
> I have noticed another UX issue, we should not reveal the document symbol, when expanding the nodes in the type hierarchy tree.

I see the same behavior with the call hierarchy. And after a quick comparison to Rider, I tend to agree that this is an UX issue. Personally I'd expect to reveal the symbol on selection. Unfortunately, expansion is equal to selection in trees. 

> We fire a selection change event when we expand the node. Perhaps, it is out of the context of this PR.

... but an interesting topic.

**Comment by simark** [comment]
One thing to look out for: make sure this plays well with restarting language servers.

If you rebase on top of #3215, you can try restarting a language server (I tested with Java, so you should be able to reproduce with that).  I get infinite recursion during dispose:

ss

**Comment by kittaakos** [comment]
> One thing to look out for: make sure this plays well with restarting language servers.

https://github.com/theia-ide/theia/pull/3148#issuecomment-428980689

> If you rebase on top of #3215,

Why to rebased? The reconnecting issue happens from the `master` branch too:  https://github.com/theia-ide/theia/issues/3145#issuecomment-429245001

**Comment by simark** [comment]
Oh sorry, I missed that comment.  Well, it gives you a more straightforward way to reproduce :).

**Comment by kittaakos** [comment]
The PR: https://github.com/theia-ide/theia/pull/3802.

**Pull request closed without merging**